endpoints/base

```diff
@@ -561,104 +569,71 @@ function process_roadblocks() {
-                        # process_bench_roadbocks to abort
-                        abort_tests_opt=1
-                    fi
-                    break
-                elif [ $a_rb -eq 0 -a $b_rb -gt 0 ]; then
-                    echo "Client-server aborted"
-                    if [ $# -gt 0 ]; then  # only if we have followers
-                        echo "Must signal collector to abort"
-                        # If ths is the last sync before running tests, we can't send an abort.
-                        # That's OK, as there's nothing left for the collectors to do, as the
-                        # abort was from 'test-start'. and collectors do not run tests.
-                        if [ $(($i+1)) -lt ${#start_syncs[*]} ]; then
-                            collector_abort_sync=${start_syncs[$(($i+1))]}
-                            echo "collector_abort_sync; $collector_abort_sync"
-                            echo "Collector will advance when sent abort signal"
-                            collector_stop_syncs=("$stop_sync" ${collector_stop_syncs[*]})
-                        fi
-                    fi
-                    no_bench=1
-                    break
-                elif [ $a_rb -eq 0 -a $b_rb -eq 0 ]; then # neither aborted, advance
-                    echo "both collector and client-server passed roadblock"
-                    collector_stop_syncs=("$stop_sync" ${collector_stop_syncs[*]})
-                    client_stop_syncs=("$stop_sync" ${client_stop_syncs[*]})
-                fi
-            done
-            echo "for loop done, collector_stop_syncs: [ ${collector_stop_syncs[*]} ]  client_stop_syncs: [ ${client_stop_syncs[*]} ]"
-            echo "client_abort_sync: $client_abort_sync   collector_abort_sync: $collector_abort_sync"
-
-            if [ $no_bench -eq 0 ]; then
-                echo "going to process bench roadblocks"
-                process_bench_roadblocks $endpoint_type $abort_tests_opt
-            else
-                if [ ! -z "$client_abort_sync" ]; then
-                    echo "going to signal client-server $client_abort_sync to abort"
-                    do_roadblock "client-server-$client_abort_sync" "follower" "abort"
-                fi
-            fi
-            echo "about to process client-server stop syncs"
-            for (( i=0; i<${#client_stop_syncs[*]}; i++ )); do
-                this_sync="${client_stop_syncs[$i]}"
-                do_roadblock "client-server-$this_sync" "follower" $client_server_script_start_timeout
-            done
-        else
-            echo "deploy aborted, going directly to cleanup"
-        fi
-        if [ ! -z "$endpoint_type" ]; then
-            endpoint_${endpoint_type}_cleanup
-        fi
-        do_roadblock "endpoint-move-data" "follower" $endpoint_deploy_timeout
+
+
+    do_roadblock "endpoint-deploy-begin" "follower" $endpoint_deploy_timeout $msg_file
+    roadblock_rc=$?
+    roadblock_exit_on_error ${roadblock_rc}
+
```

Comment:
So one of the things I tried to do here is to actually unroll the loops that were present.  The way it was originally written the loops would iterate over a set of synchronization points (ex. `A`, `B`, & `C`) and create `A-start`, `B-start`, and `C-start` while also creating (in reverse order) `C-stop`, `B-stop`, and `A-stop`.  The problem with this was when you got a roadblock failure during a run, lets say at `B-stop`, you would go look at the code and there was no way to easily find where the `B-stop` synchronization point was because it didn't actually exist as that name anywhere in the code -- the name was pieced together as the loop was iterated over and various conditional statements were evaluated.  You would eventually be able to locate the problem if you knew the code really well and could follow all the logic but it was a very high barrier to entry.

To make it even more complicated this was happening in 3 different scripts (`rickshaw-run`, `endpoints/base`, and `engine/client-server-script`) while being implemented in two different languages (`rickshaw-run` is in perl, `endpoints/base` and `engine/client-server-script` are in bash shell scripts) and the conditionals were different between all 3 scripts because each one is doing slightly different things depending on it's role in the current synchronization point.

The big change here, other than a renaming of the synchronization points and the creation of some new ones is that that loop logic has been eliminated.  This allows a user to go to one of the scripts with the name of a synchronization point they failed at and easily search for and find that point in the code.  They can then see exactly what is happening at that synchronization point and try to pinpoint what the problem is.

The other thing that was happening in the loops is there was a lot of complex logic being used to track progress so that it would know what had already happened so that it would know what to do in the case of a failure -- ie. since `A-start` happened we need to create a `A-stop`.  This dynamic behavior resulted in us not doing proper cleanup and continuation in the face of errors which is one of the reasons why a failed run would exit uncleanly.  The new code tries to continue all the way to completion as much as possible once a certain point is reached -- namely the test phase.  The tracking that was being done doesn't need to be done if we are always trying to run to completion by processing all synchronization points every time, it just needs to know if there is some type of error (ie. an abort or a quit) and how to properly handle that (ie. skip running tests when a quit is signaled but continue making progress).  The other thing that happens is that if a timeout is observed, the only way to successfully make progress through the remaining roadblocks is to remove the follower(s) that did not check in to a roadblock from the future roadblocks so that they do not continue to cause timeouts.

You can probably sense a theme of trying to make it easier to follow the code for trouble shooting reasons, so I also tried to make the code as similar as possible across the 3 different lines of execution.  I changed some of the function names so that they are the same and tried to create commonly named and functional new ones as much as possible.  In some cases the implementation details are slightly different but the overall intent is supposed to be the same and therefore make it easier to follow -- at least that is the hope/intent.  I've also tried to move chunks of the logic that were formerly in the loops into functions so that it can easily be called without negatively impacting the ability to follow the flow of exection.

You are accurate in saying that this is very hard to follow due to it's size, but the unfortunate reality is that since we have these 3 different scripts that synchronize with each other as they run and it was originally done with the highly integrated loops that it is very difficult, if not impossible, to break up the changes into smaller more incremental pieces.  I'm hoping that going forward that is not the case from now on.  For example, it should be relatively straight forward to insert one or more new synchronization points at any point in the process now across all 3 scripts with minimal work, I think.